File: frontend/src/api/ingestion.ts
export const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://127.0.0.1:8000";

export type JobLifecycleStatus = "pending" | "running" | "succeeded" | "failed";
export type StageStatus = "pending" | "running" | "succeeded" | "failed" | "skipped";
export type ArtifactType =
  | "public_metadata"
  | "transcript"
  | "media_retrieval"
  | "screenshot"
  | "claims"
  | "research_basis";
export type ResearchBasisStatus =
  | "source_candidates_found"
  | "no_research_source_found"
  | "opinion_or_unratable"
  | "needs_manual_review";

export type IngestionStage = {
  name: string;
  status: StageStatus;
  message: string;
  entity_uuid?: string | null;
};

export type ArtifactStatus = {
  artifact_type: ArtifactType;
  status: StageStatus;
  label: string;
  entity_uuid?: string | null;
  message: string;
  details: Record<string, unknown>;
};

export type TranscriptSegment = {
  uuid: string;
  start_seconds?: number | null;
  end_seconds?: number | null;
  text: string;
};

export type TranscriptArtifact = {
  transcript_uuid: string;
  video_uuid: string;
  provenance: {
    method: string;
    source_url: string;
    provider?: string | null;
    quality_note?: string | null;
    failure_reason?: string | null;
  };
  segments: TranscriptSegment[];
  plain_text: string;
  vault_path?: string | null;
};

export type ScreenshotArtifact = {
  screenshot_uuid: string;
  video_uuid: string;
  timestamp_seconds?: number | null;
  vault_path: string;
  asset_url?: string | null;
  source_clue: boolean;
  source_clue_text?: string | null;
  claim_uuids: string[];
};

export type ExtractedClaim = {
  uuid: string;
  source_video_uuid: string;
  source_transcript_uuid: string;
  timestamp_start_seconds?: number | null;
  timestamp_end_seconds?: number | null;
  transcript_excerpt: string;
  claim_text: string;
  screenshot_uuids: string[];
  extraction_confidence?: number | null;
  evidence_status: "pending";
  source_candidate_count: number;
};

export type ResearchBasisCandidate = {
  uuid: string;
  candidate_type: string;
  value: string;
  source: string;
  source_uuid?: string | null;
  confidence?: number | null;
};

export type ResearchBasisTriage = {
  status: ResearchBasisStatus;
  candidates: ResearchBasisCandidate[];
  reason: string;
  candidate_count: number;
};

export type IngestionJob = {
  job_uuid: string;
  source_kind: string;
  source_url: string;
  status: JobLifecycleStatus;
  current_operation: string;
  stages: IngestionStage[];
  artifacts: ArtifactStatus[];
  created_at: string;
  updated_at: string;
  error_message?: string | null;
  video_uuid?: string | null;
  public_metadata?: Record<string, unknown>;
  transcript_artifact?: TranscriptArtifact | null;
  screenshots?: ScreenshotArtifact[];
  claims?: ExtractedClaim[];
  research_basis?: ResearchBasisTriage;
};

export function displayLifecycleStatus(status: JobLifecycleStatus) {
  return status === "succeeded" ? "complete" : status;
}

async function parseResponse(response: Response): Promise<IngestionJob> {
  if (!response.ok) {
    const payload = await response.json().catch(() => ({}));
    throw new Error(typeof payload.detail === "string" ? payload.detail : "ingestion_request_failed");
  }

  return response.json() as Promise<IngestionJob>;
}

export async function submitTikTokUrl(url: string): Promise<IngestionJob> {
  return parseResponse(
    await fetch(`${API_BASE_URL}/ingestion/tiktok`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ url }),
    }),
  );
}

export async function submitPastedTranscript(
  sourceUrl: string,
  transcript: string,
): Promise<IngestionJob> {
  return parseResponse(
    await fetch(`${API_BASE_URL}/ingestion/fixtures/transcript`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ source_url: sourceUrl, transcript }),
    }),
  );
}

export async function uploadVideoFile(file: File, transcript?: string): Promise<IngestionJob> {
  const form = new FormData();
  form.append("file", file);
  if (transcript) {
    form.append("transcript", transcript);
  }

  return parseResponse(
    await fetch(`${API_BASE_URL}/ingestion/videos/upload`, {
      method: "POST",
      body: form,
    }),
  );
}

export async function fetchIngestionJob(jobUuid: string): Promise<IngestionJob> {
  return parseResponse(await fetch(`${API_BASE_URL}/ingestion/jobs/${jobUuid}`));
}

export function resolveApiAssetUrl(assetUrl?: string | null): string | null {
  if (!assetUrl) {
    return null;
  }

  if (/^https?:\/\//i.test(assetUrl) || assetUrl.startsWith("data:")) {
    return assetUrl;
  }

  return `${API_BASE_URL}${assetUrl.startsWith("/") ? "" : "/"}${assetUrl}`;
}

File: backend/app/schemas/ingestion.py
from datetime import datetime, timezone
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


class JobLifecycleStatus(str, Enum):
    pending = "pending"
    running = "running"
    succeeded = "succeeded"
    failed = "failed"


class StageStatus(str, Enum):
    pending = "pending"
    running = "running"
    succeeded = "succeeded"
    failed = "failed"
    skipped = "skipped"


class SourceKind(str, Enum):
    tiktok_url = "tiktok_url"
    uploaded_video = "uploaded_video"
    fixture_transcript = "fixture_transcript"


class IngestionStageName(str, Enum):
    validate_url = "validate_url"
    read_public_metadata = "read_public_metadata"
    build_transcript = "build_transcript"
    capture_source_clues = "capture_source_clues"
    extract_claims = "extract_claims"
    triage_research_basis = "triage_research_basis"
    write_owned_artifacts = "write_owned_artifacts"


class ArtifactType(str, Enum):
    public_metadata = "public_metadata"
    transcript = "transcript"
    media_retrieval = "media_retrieval"
    screenshot = "screenshot"
    claims = "claims"
    research_basis = "research_basis"


class TranscriptRetrievalMethod(str, Enum):
    public_caption = "public_caption"
    subtitle_file = "subtitle_file"
    transcription = "transcription"
    pasted = "pasted"
    fixture = "fixture"


class ResearchBasisStatus(str, Enum):
    source_candidates_found = "source_candidates_found"
    no_research_source_found = "no_research_source_found"
    opinion_or_unratable = "opinion_or_unratable"
    needs_manual_review = "needs_manual_review"


class IngestionStage(BaseModel):
    name: IngestionStageName
    status: StageStatus = StageStatus.pending
    message: str = ""
    entity_uuid: UUID | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None


class ArtifactStatus(BaseModel):
    artifact_type: ArtifactType
    status: StageStatus
    label: str
    entity_uuid: UUID | None = None
    message: str = ""
    details: dict[str, Any] = Field(default_factory=dict)


class TranscriptSegment(BaseModel):
    uuid: UUID = Field(default_factory=uuid4)
    start_seconds: float | None = None
    end_seconds: float | None = None
    text: str


class TranscriptProvenance(BaseModel):
    method: TranscriptRetrievalMethod
    source_url: str
    source_artifact_uuid: UUID | None = None
    provider: str | None = None
    confidence: float | None = None
    quality_note: str | None = None
    failure_reason: str | None = None


class TranscriptArtifact(BaseModel):
    transcript_uuid: UUID = Field(default_factory=uuid4)
    video_uuid: UUID
    provenance: TranscriptProvenance
    segments: list[TranscriptSegment] = Field(default_factory=list)
    plain_text: str
    vault_path: str | None = None


class ScreenshotArtifact(BaseModel):
    screenshot_uuid: UUID = Field(default_factory=uuid4)
    video_uuid: UUID
    timestamp_seconds: float | None = None
    vault_path: str
    asset_url: str | None = None
    source_clue: bool = False
    source_clue_text: str | None = None
    claim_uuids: list[UUID] = Field(default_factory=list)


class ResearchBasisCandidate(BaseModel):
    uuid: UUID = Field(default_factory=uuid4)
    candidate_type: str
    value: str
    source: str
    source_uuid: UUID | None = None
    confidence: float | None = None


class ResearchBasisTriage(BaseModel):
    status: ResearchBasisStatus
    candidates: list[ResearchBasisCandidate] = Field(default_factory=list)
    reason: str

    @property
    def candidate_count(self) -> int:
        return len(self.candidates)


class IngestionJob(BaseModel):
    job_uuid: UUID = Field(default_factory=uuid4)
    source_kind: SourceKind
    source_url: str
    status: JobLifecycleStatus = JobLifecycleStatus.pending
    current_operation: str = "Waiting to start ingestion."
    stages: list[IngestionStage] = Field(default_factory=list)
    artifacts: list[ArtifactStatus] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)
    error_message: str | None = None
    video_uuid: UUID | None = None
